---
title: "scNetViz: Use case #1"
author: "Krishna Choudhary, Alex Pico"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

This use case will demonstrate analyzing and visualizing single cell datasets from within Cytoscape to generate network views of clusters of differentiated genes per cell population. In order to control Cytoscape from R, we'll need to install the RCy3 pacakge:

# Installation
```{r}
if(!"RCy3" %in% installed.packages()){
    install.packages("BiocManager")
    BiocManager::install("RCy3")
}
library(RCy3)
```

The whole point of RCy3 is to connect with Cytoscape, so you will need to install and launch Cytoscape: 
    
* Download the latest Cytoscape from http://www.cytoscape.org/download.php
* Complete installation wizard
* Launch Cytoscape 

```{r}
cytoscapePing()
```

For this use case, you'll also need two Cytoscape apps:
 * [scNetViz app](http://apps.cytoscape.org/apps/scnetviz) to work with single cell datasets in Cytoscape
 * [stringApp](https://apps.cytoscape.org/apps/stringapp) to perform enrichment analysis using the STRING webservice

```{r}
installApp('scNetViz')  
installApp('STRINGapp')  
```

In this example, we will browse a single cell expression atlas, explore a particular dataset, perform differential expression analysis based on provided categories, generate networks from the top genes from each category, and functionally characterize and visualize the networks. 

# Load single cell dataset
Use the accession number of single-cell experiment to pull data from the  [Single-Cell Experiment Atlas](https://www.ebi.ac.uk/gxa/sc/experiments) of EMBL-EBI.


```{r}
#Load data with Accession number E-GEOD-81383
RCy3::commandsRun('scnetviz load gxa experiment accession=E-GEOD-81383')
```

This loads the data and opens an experiment table with three tabs, named _TPM_, _Categories_, and _DiffExp_. 

# Perform differential expression analysis
Run differential expression analysis for the row with `true` value of `sel.K` (default). 
```{r}
RCy3::commandsRun('scnetviz calculate diffexp accession=E-GEOD-81383')
```

# Generate networks
Fetch protein-protein interaction networks from the [STRING](https://string-db.org/) database.

```{r}
RCy3::commandsRun('scnetviz create network accession=E-GEOD-81383')
```

# Check resulting networks
Check the networks available in the current Cytoscape session.
```{r}
RCy3::commandsRun('network list')
```

# Perform functional enrichment analysis
Perform functional enrichment analysis for the network selected in the current session using the _stringApp_. 

```{r}
net.enr <- RCy3::commandsRun('string retrieve enrichment allNetSpecies=Homo sapiens')
RCy3::commandsRun('string show enrichment') 
```

Visit the [scNetViz website](https://www.cgl.ucsf.edu/cytoscape/scNetViz/#networks) for more information.






